File: frontend/src/types/case.ts
/**
 * Case (scenario) types for admin and trainee flows.
 *
 * @remarks
 * {@link CaseDTO} matches backend JSON (`snake_case`). {@link Case} is the camelCase domain type used in the UI.
 */

/** Wire shape for a training case as returned or accepted by the API. */
export interface CaseDTO {
  id: number
  title: string
  description?: string | null
  script: string
  objectives?: string | null
  difficulty_level?: 'beginner' | 'intermediate' | 'advanced' | string
  category?: string | null
  patient_background?: string | null
  expected_spikes_flow?: string | null
  evaluator_plugin?: string | null
  patient_model_plugin?: string | null
  metrics_plugins?: string[] | null
  created_at: string
  updated_at: string
}

/** Paginated list of cases from the API. */
export interface CaseListDTO {
  cases: CaseDTO[]
  total: number
}

/** Case entity in camelCase for components and forms. */
export interface Case {
  id: number
  title: string
  description?: string
  script: string
  objectives?: string
  difficultyLevel?: string
  category?: string
  patientBackground?: string
  expectedSpikesFlow?: string
  evaluatorPlugin?: string | null
  patientModelPlugin?: string | null
  metricsPlugins?: string[] | null
  createdAt: string
  updatedAt: string
}

File: backend/src/domain/entities/case.py
"""Case entity model."""

from sqlalchemy import Column, Integer, String, Text
from sqlalchemy.orm import relationship

from core.time import utc_now
from db.base import Base
from db.types import UTCDateTimeType


class Case(Base):
    """Case model for simulation scenarios."""
    
    __tablename__ = "cases"
    
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(Text)
    script = Column(Text, nullable=False)  # Detailed scenario script
    objectives = Column(Text)  # JSON or text describing learning objectives
    difficulty_level = Column(String)  # beginner, intermediate, advanced
    category = Column(String)  # e.g., oncology, pediatrics, etc.
    patient_background = Column(Text)  # Patient history and context
    expected_spikes_flow = Column(Text)  # JSON describing expected SPIKES progression
    evaluator_plugin = Column(String, nullable=True)  # Optional case-level evaluator override (registry key)
    patient_model_plugin = Column(String, nullable=True)  # Optional case-level patient model override (registry key)
    metrics_plugins = Column(Text, nullable=True)  # JSON array of metrics plugin names, e.g. ["plugin.name:Class"]
    created_at = Column(UTCDateTimeType(), default=utc_now)
    updated_at = Column(UTCDateTimeType(), default=utc_now, onupdate=utc_now)
    
    # Relationships
    sessions = relationship("Session", back_populates="case", cascade="all, delete-orphan")
    
    def __repr__(self) -> str:
        return f"<Case(id={self.id}, title={self.title}, difficulty={self.difficulty_level})>"

